psycopg_pool/psycopg_pool/pool.py

```diff
@@ -425,11 +425,12 @@ def close(self, timeout: float = 5.0) -> None:
             waiting = list(self._waiting)
             self._waiting.clear()
             connections = list(self._pool)
-            self._pool.clear()
 
-        # Now that the flag _closed is set, getconn will fail immediately,
-        # putconn will just close the returned connection.
-        self._stop_workers(waiting, connections, timeout)
+            # Now that the flag _closed is set, getconn will fail immediately,
+            # putconn will just close the returned connection.
+            self._stop_workers(waiting, connections, timeout)
+
+            self._pool.clear()
```

Comment:
As you, I have no idea how you can get it the situation in which there are still connections in the pool. Maybe we call `_stop_workers()` in a moment in which a worker is busy creating a connection, and it will receive the `StopWorker` task only after it has added a new connection. Therefore I see how calling `_pool.clear()` afterwards makes something.

One thing that is note done right here is the `_pool.clear()` which doesn't close a connection eventually left there, which may result in a warning.

Looking at the `_stop_worker()`, I find the `connections` parameter extremely weird. It doesn't seem its responsibility to do that. Can you please try to refactor this code by removing the `connections` parameter from `_stop_workers` and then, as you did here, clear the pool, but closing the connection, after `_stop_worker()` call? Something like:

```python
        with self._lock:
            # [snip]
            self._waiting.clear()

            self._stop_workers(waiting, timeout)

            for conn in self._pool:
                conn.close()
            self._pool.clear()

        # Now that the flag _closed is set, getconn will fail immediately,
        # putconn will just close the returned connection.
```

Reply:
I see that `_stop_workers()` is also called by `__del__()`, but with no connection list, and also that it takes on itself the responsibility of closing the waiting clients... I don't know if this is really the best organisation of this code: feel free to propose cleanup refactoring if you see any obvious one.